drivers/i2c: stm32: Use dts for i2c clock source

gautierg-st opened the pull request:
This PR adds support of the I2C clock source for STM32 in the dts file.

Fixes #52675

gautierg-st committed:
dts: arm: st: Add default I2C clock source for STM32F0 & F3

Define SYSCLK as the default I2C source clock for I2C1 on STM32F0x
and all I2Cx on STM32F3x.

On most series, the default I2C clock source (when it exists) is PCLK.
This clock does not exist as I2C clock source on FO & F3 and the default
one is HSI. Since HSI is not necessarily enabled we explicitly set it
to SYSCLK instead.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: i2c: Use dts to determine i2c clock source for STM32

Add the possibility to use the clock parameter in dts to set the clock
source of I2C devices for all STM32 MCUs.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on include/zephyr/dt-bindings/clock/stm32f4_clock.h:
```diff
@@ -68,4 +71,27 @@
 /** BDCR devices */
 #define RTC_SEL(val)		STM32_CLOCK(val, 3, 8, BDCR_REG)
 
+/* Following definitions are for STM32F7X only */
```
Please introduce a dedicated stm32f7_clocks.h

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -179,7 +179,8 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x40005400 0x400>;
-			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00200000>;
```
Please provide more comments in the commit message why we are defining default clock sources (which is not a common practice). According to existing code, I can see:
```
	 * STM32F0/3 I2C's independent clock source supports only
	 * HSI and SYSCLK, not APB1. We force I2C clock source to SYSCLK.
	 * I2C2 on STM32F0 uses APB1 clock as I2C clock source
```

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -230,39 +230,20 @@ static int i2c_stm32_init(const struct device *dev)
-		LL_RCC_SetI2CClockSource(LL_RCC_I2C3_CLKSOURCE_SYSCLK);
-		break;
-#endif
+#if (DT_NUM_CLOCKS(DT_NODELABEL(name)) > 1)
```
Contrary to `STM32_I2C_INIT()` case, `name` is undefined, this will likely not produce the expected result.

gautierg-st committed:
include: zephyr: dt-bindings: clock: Add separate clock file for STM32F7

Add a new dedicated clock control header file for the STM32F7 series.
It is copied from the STM32F4 file with added clock defines and helpers
to select peripheral clock sources that are specific to F7 series.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

dt-bindings f7

gautierg-st committed:
dts: arm: st: f7: Use new dedicated clock file for STM32F7

Include the new clock file dedicated for STM32F7 instead of the F4 one
previously used.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on include/zephyr/dt-bindings/clock/stm32f7_clock.h:
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2022 Linaro Limited
```
Fix copyright

erwango reviewed (approved):
LGTM
